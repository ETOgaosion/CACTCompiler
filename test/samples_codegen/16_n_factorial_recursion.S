iobuff:
	.zero 80
.text
.align 1
.globl	factorial
.type	factorial, @function
factorial:
	sd	ra, -8(sp)
	sw	a0, -12(sp)
	lw	t0, -12(sp)
	addi	t1, zero, 1
	xor	t2, t0, t1
	seqz	t3, t2
	sb	t3, -28(sp)
	lb	t0, -28(sp)
	beqz	t0, L_0_factorial
	lw	a0, -12(sp)
	ld	ra, -8(sp)
	ret
	lla	t0, L_1_factorial
	jr	t0
L_0_factorial:
	lw	t0, -12(sp)
	addi	t1, zero, 1
	subw	t2, t0, t1
	sw	t2, -24(sp)
	lw	a0, -24(sp)
	addi	sp, sp, -32
	call	factorial
	addi	sp, sp, 32
	sw	a0, -20(sp)
	lw	t0, -12(sp)
	lw	t1, -20(sp)
	mulw	t2, t0, t1
	sw	t2, -16(sp)
	lw	a0, -16(sp)
	ld	ra, -8(sp)
	ret
L_1_factorial:
	ld	ra, -8(sp)
	ret
.text
.align 1
.globl	main
.type	main, @function
main:
	sd	ra, -8(sp)
	mv	t0, zero
	sw	t0, -16(sp)
	mv	t0, zero
	sw	t0, -12(sp)
	addi	sp, sp, -32
	call	get_int
	addi	sp, sp, 32
	sw	a0, -20(sp)
	lw	t0, -20(sp)
	sw	t0, -16(sp)
	lw	t0, -16(sp)
	addi	t1, zero, 1
	slt	t2, t1, t0
	sb	t2, -24(sp)
	lb	t0, -24(sp)
	beqz	t0, L_0_main
	lw	a0, -16(sp)
	addi	sp, sp, -32
	call	factorial
	addi	sp, sp, 32
	sw	a0, -28(sp)
	lw	t0, -28(sp)
	sw	t0, -12(sp)
	lla	t0, L_1_main
	jr	t0
L_0_main:
	addi	t0, zero, 1
	sw	t0, -12(sp)
L_1_main:
	lw	a0, -12(sp)
	addi	sp, sp, -32
	call	print_int
	addi	sp, sp, 32
	mv	a0, zero
	ld	ra, -8(sp)
	ret
