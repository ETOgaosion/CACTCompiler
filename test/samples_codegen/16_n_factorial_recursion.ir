============ global var ================
============ imm value ================
============ functions ================
----------- funcDef -----------
factorial(2) => 2:
----------- paramVars -----------
n := symbol type: 0; data type: 2
----------- localVars -----------
----------- tempVars ------------
t_3:= symbol type: TEMP VAR; data type: 2
t_2:= symbol type: TEMP VAR; data type: 2
t_1:= symbol type: TEMP VAR; data type: 2
t_0:= symbol type: TEMP VAR; data type: 1
----------- Codes ------------
factorial:
	n =  param 0;
	t_0 = n == 1
	beqz t_0 L_0_factorial
	return ;
	goto L_1_factorial
L_0_factorial:
	t_1 = n - 1;
	addparam t_1;
	call factorial; 
	t_2 =  return ;
	t_3 = n * t_2;
	return ;
L_1_factorial:
	return ;

----------- funcDef -----------
main() => 2:
----------- paramVars -----------
----------- localVars -----------
result := symbol type: 1; data type: 2; initValue: 0

n := symbol type: 1; data type: 2; initValue: 0

----------- tempVars ------------
t_0:= symbol type: TEMP VAR; data type: 2
t_2:= symbol type: TEMP VAR; data type: 1
t_3:= symbol type: TEMP VAR; data type: 2
----------- Codes ------------
main:
	n = 0;
	result = 0;
	call get_int; 
	t_0 =  return ;
	t_1 <= n;
	t_1 = t_0;
	t_2 = t_1 > 1
	beqz t_2 L_0_main
	addparam t_1;
	call factorial; 
	t_3 =  return ;
	t_4 <= result;
	t_4 = t_3;
	goto L_1_main
L_0_main:
	t_5 <= result;
	t_5 = 1;
L_1_main:
	addparam t_5;
	call print_int; 
	return 0;

