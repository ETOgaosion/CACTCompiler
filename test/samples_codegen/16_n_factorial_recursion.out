bbl loader
please enter an int number:
1
